### PR TITLE
FINERACT-1724: Datatables query API fix - invalid filter query

### PR DESCRIPTION
**Pull request by abraham-menyhart**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java [hunk 1]
```diff
@@ -19,6 +19,8 @@
 package org.apache.fineract.infrastructure.dataqueries.service;
 
 import static java.util.Arrays.asList;
+import static java.util.Locale.ENGLISH;
+import static org.apache.commons.lang3.time.DateFormatUtils.ISO_8601_EXTENDED_DATE_FORMAT;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Please rather use `DateTimeFormatter.ISO_LOCAL_DATE`.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java [hunk 2]
```diff
@@ -265,37 +273,37 @@ private void extractResults(SqlRowSet rowSet, String[] resultColumnNames, List<J
     }
 
     @NotNull
-    private SqlRowSet callFilteredMysql(String sql, String valueFilter, String filterColumnType) {
+    private SqlRowSet callFilteredMysql(String sql, Object valueFilter, String filterColumnType) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Please use specific type!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java [hunk 3]
```diff
@@ -265,37 +273,37 @@ private void extractResults(SqlRowSet rowSet, String[] resultColumnNames, List<J
...
         Object finalValueFilter = valueFilter;
         SqlRowSet rowSet;
         if (BIT_TYPE.equalsIgnoreCase(filterColumnType)) {
             int[] argType = new int[1];
             argType[0] = Types.BIT;
-            finalValueFilter = BooleanUtils.toString(BooleanUtils.toBooleanObject(valueFilter), "1", "0", "null");
+            finalValueFilter = BooleanUtils.toString(BooleanUtils.toBooleanObject(valueFilter.toString()), "1", "0", "null");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Please dont use .toString()! 90% of the cases it is bad pattern!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 1 above)
Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: I would, but the input of the validateColumn method is a string... And I didn't want to refactor the validateColumn method to accept  DateTimeFormatter. You ask too much @adamsaghy ! :moneybag: But I'll see want I can do :wink:

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 2 above)
Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Same old validateColumn method, that I have to refactor then :cry:

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 3 above)
Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: I was waiting for this comment! I totally agree with you, and I'm already thinking about how to fix it!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Its okay, leave it as is

**Comment by adamsaghy** [comment]
@abraham-menyhart Kindly asking you to check the failing tests!

File: fineract-provider/src/test/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImplTest.java [hunk 4]
```diff
@@ -81,30 +85,56 @@ public void testSqlInjectionCaughtQueryDataTable2() {
...
+        when(jdbcTemplate.queryForRowSet(eq("select rc1,rc2 from table where cf1 = ?"), any(Object.class))).thenReturn(sqlRS);
         when(sqlRS.next()).thenReturn(true).thenReturn(false);
         when(sqlRS.getObject(ArgumentMatchers.anyString())).thenReturn("value1").thenReturn("value2");
         when(sqlGenerator.escape(ArgumentMatchers.anyString())).thenReturn("rc1").thenReturn("rc2").thenReturn("table").thenReturn("cf1");
         when(databaseTypeResolver.isPostgreSQL()).thenReturn(true);
 
-        ResultsetColumnHeaderData cf1 = ResultsetColumnHeaderData.detailed("cf1", "text", 10L, false, false, null, null, false, false);
-        ResultsetColumnHeaderData rc1 = ResultsetColumnHeaderData.detailed("rc1", "text", 10L, false, false, null, null, false, false);
-        ResultsetColumnHeaderData rc2 = ResultsetColumnHeaderData.detailed("rc2", "text", 10L, false, false, null, null, false, false);
+        ResultsetColumnHeaderData cf1 = ResultsetColumnHeaderData.detailed("cf1", "text", 10L, false, false, emptyList(), null, false,
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImplTest.java:94]: why change the null to emptyList?

**Review by adamsaghy — APPROVED** [review]
LGTM

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**